File: side-project/studyolle_redo/src/main/java/com/redo/studyolle/modules/domain/entity/Account.java
package com.redo.studyolle.modules.domain.entity;

import lombok.*;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

@Entity
@Getter
@Setter // modelMapper
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Account {
    @Id
    @GeneratedValue
    private Long id;

    @Column(unique = true)
    private String email;

    @Column(unique = true)
    private String nickname;

    private String password;

    private boolean emailVerified;

    private String emailCheckToken;

    private LocalDateTime emailCheckTokenGeneratedAt;

    private LocalDateTime joinedAt;

    private String bio;

    private String url;

    private String occupation;

    private String location;

    @Lob
    @Basic(fetch = FetchType.EAGER)
    private String profileImage;

    private boolean studyCreatedByEmail;

    private boolean studyCreatedByWeb = true;

    private boolean studyEnrollmentResultByEmail;

    private boolean studyEnrollmentResultByWeb = true;

    private boolean studyUpdatedByEmail;

    private boolean studyUpdatedByWeb = true;

    /*
      create table account_tags (
        account_id int8 not null,
        tags_id int8 not null,
        primary key (account_id, tags_id)
    )
     */
    @ManyToMany
    private Set<Tag> tags = new HashSet<>();

    @ManyToMany
    private Set<Zone> zones = new HashSet<>();

    /**
     * 이메일 토큰 생성하여 저장
     */
    public void generateEmailCheckToken() {
        this.emailCheckToken = UUID.randomUUID().toString();
        this.emailCheckTokenGeneratedAt = LocalDateTime.now();
    }

    /**
     * 토큰 유효성 체크
     * @param token
     * @return
     */
    public boolean isValidToken(String token) {
        return this.emailCheckToken.equals(token);
    }

    /**
     * 이메일 토큰 인증 정보 저장
     */
    public void completeSignUp() {
        this.emailVerified = true;
        this.joinedAt = LocalDateTime.now();
    }

    /**
     * 이메일 로그인 발송 1시간마다 1회씩 가능
     * @return
     */
    public boolean canSendConfirmEmail() {
        return this.emailCheckTokenGeneratedAt.isBefore(LocalDateTime.now().minusHours(1));
    }
}

File: src/main/kotlin/com/studyolle/domain/Account.kt
package com.studyolle.domain

import java.time.LocalDateTime
import javax.persistence.*

@Entity
class Account (
    @Id
    @GeneratedValue
    val id: Long? = null,

    /* 중복 불가능 */
    @Column(unique = true)
    val email: String? = null,

    @Column(unique = true)
    val nickname: String? = null,
    val password: String? = null,

    /* 이메일 검증 */
    var emailVerified : Boolean = false,
    var emailCheckToken: String? = null,
    var emailCheckTokenGeneratedAt: LocalDateTime? = null,
    var joinedAt : LocalDateTime? = null,
    val bio : String? = null, // 자기소개
    val url: String? = null,
    val occupation : String? = null,
    val location : String? = null,

    /* CLOB */
    @Lob
    @Basic(fetch = FetchType.EAGER)
    /* 즉시로딩 */
    val profileImage : String? = null,

    val studyCreatedByEmail : Boolean = false, // 스터디 생성 여부를 이메일로 받을것인가?
    val studyCreatedByWeb : Boolean = false, // 스터디 생성 여부를 웹으로 받을것인가?
    val studyEnrollmentResultByEmail : Boolean = false, // 스터디 모임 가입신청 결과를 이메일로 받을것인가?
    val studyEnrollmentResultByWeb : Boolean = false, // 스터디 모임 가입신청 결과를 웹으로 받을것인가?
    val studyUpdatedByEmail : Boolean = false, // 스터디 변경 정보를 이메일로 받을것인가?
    val studyUpdatedByWeb : Boolean = false, // 스터디 변경 정보를 웹으로 받을것인가?

    /* account_tags 테이블 생성 */
    @ManyToMany
    val tags: Set<Tag> = HashSet(),

    /* account_zones 테이블 생성 */
    @ManyToMany
    val zones: Set<Zone> = HashSet(),
) {
}
